Access Control - Group admin page

MarieVerdonck opened the pull request:
This PR adds the admin access control page for Groups.

Functionalities:

- (Read) Initial page load gives you a pageable list of all Groups. This list is searchable. Per Group ID, Name, #Members (epersons and subgroups) is shown.
- (Create) A new Group can be created, the name (and optional description) must first be submitted before subgroups/epeople can be added.
- (Delete) The delete of a group is not yet implemented in the backend, so this button currently just shows a notification
- (Update) Changing a group’s name / description is not yet implemented in the backend, so submitting the edit form just shows a notification
- (Update Members): Members (people and groups) can be added/deleted. For this there is a searchable list of both on the edit group page. The edit group page is accessible via the access-control group list, or via the url containing the group’s uuid (/admin/access-control/groups/). Initially at edit group page load the list of subgroups and epersons that group has, is shown, but again a search is possible.

This is dependent on backend PR: https://github.com/DSpace/DSpace/pull/2686 and frontend PR for the EPeople access control admin page: https://github.com/DSpace/dspace-angular/pull/609

tdonohue commented:
NOTE: This PR will need minor updates as the subgroups endpoint was renamed to `/groups/[uuid]/subgroups` (previously was `/groups/[uuid]/groups`).  

Additionally, as a sidenote, this PR looks to be failing code coverage tests as it looks to be missing specs for all new components/classes.

MarieVerdonck commented:
- rebased from upstream/master
- Group.groups to Group.subgroups
- Added notification error when group creation failed
- Specific notification error when group creation failed because of name already in use (no exact search on name possible so will give error if there is a search result on the name used for the creation attempt And if the attempt failed
- Test spec files added for group data service, groups registry, groups reducer, add new group form, subgroups list and members list

tdonohue commented on src/app/core/cache/object-cache.service.ts:
```diff
@@ -276,6 +276,8 @@ export class ObjectCacheService {
    */
   public addPatch(selfLink: string, patch: Operation[]) {
+    console.log('selfLink addPatch', selfLink)
+    console.log('patch addPatch', patch)
```
I suspect these log statements were just for debugging and can be removed now.

MarieVerdonck commented:
@tdonue:
* Added check with Group.permanent on whether or not to show delete button
* Disabled groupName input field if group that is being edited is permanent
* Cancel edit group button sends you back to group registry now
* In group edit, cache only gets cleared of that groups subgroups and epersons cache (at member delete/add) instead of cleared of all groups info
* Separated search groups result list and list of subgroups in edit group page and eperson members of edit group page
* Fixed pagination after search issues in groups registry, group edit page (groups search, subgroups, epeople search and members)
* Applied same fixes to epeople registry
* Browse All button added to all group/eperson searches to browse all groups/epeople
* Fixed immediately being able to add members after group creation in edit group page
* Fixed hover i18n tags on groups & epeople registry page => Missing tags
* Added message if trying to add current group (that is being edited) as own subgroup
* Removed possible add/delete button from current group being edited in subgroups list, replaced with "Current groups" message
* Fixed tests after changes

MarieVerdonck commented:
@tdonohue: 
- Messages altered
- I’m seeing the same issue but not consistently and also sometimes on subgroups and also on the subgroups/epeople delete with the icons not immediately updating; so maybe this is a caching issue? Adding members happens like so:
    - When eperson1 is added to activeGroup => eperson1.groups and activeGroup.epersons lists get cleared from cache => list of epeople members get populated again with new call the backend to activeGroup.epersons
    - When subgroup1 gets added to activeGroup => activeGroup.subgroups gets cleared from cache => list of subgroups gets populated again with new call to the backend to activeGroup.subgroups
    - The same when deleting either
=> Changed it so that the clearing of the cache gets done, without having to wait for (successful) response first, which seemed to have fixed it for me I think (tested with yarn run start)
- Fixed so after refresh the activeGroup should get set again after refresh of existing group edit page (tested with yarn run start and seems to work okay)

tdonohue commented:
Merging as-is, as @artlowel and I agree this is a good enough initial implementation. Further improvements will be made in a follow-up PR, as documented in ticket #647   Thanks @MarieVerdonck !
